File: python/src/taskulus/cache.py
"""Index cache utilities for Taskulus."""

from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict, List, Optional

from taskulus.index import IssueIndex
from taskulus.models import IssueData


@dataclass
class IndexCache:
    """Serialized index cache representation."""

    version: int
    built_at: datetime
    file_mtimes: Dict[str, float]
    issues: List[IssueData]
    reverse_deps: Dict[str, List[str]]


def collect_issue_file_mtimes(issues_directory: Path) -> Dict[str, float]:
    """Collect file modification times for issues.

    :param issues_directory: Directory containing issue files.
    :type issues_directory: Path
    :return: Mapping of filename to mtime.
    :rtype: Dict[str, float]
    """
    return {path.name: path.stat().st_mtime for path in issues_directory.glob("*.json")}


def load_cache_if_valid(
    cache_path: Path, issues_directory: Path
) -> Optional[IssueIndex]:
    """Load cached index if the cache is valid.

    :param cache_path: Path to cache file.
    :type cache_path: Path
    :param issues_directory: Directory containing issue files.
    :type issues_directory: Path
    :return: IssueIndex if cache is valid, otherwise None.
    :rtype: Optional[IssueIndex]
    """
    if not cache_path.exists():
        return None

    payload = json.loads(cache_path.read_text(encoding="utf-8"))
    file_mtimes = payload.get("file_mtimes", {})
    current_mtimes = collect_issue_file_mtimes(issues_directory)
    if file_mtimes != current_mtimes:
        return None

    issues = [IssueData.model_validate(item) for item in payload.get("issues", [])]
    reverse_deps = payload.get("reverse_deps", {})
    return build_index_from_cache(issues, reverse_deps)


def write_cache(
    index: IssueIndex, cache_path: Path, file_mtimes: Dict[str, float]
) -> None:
    """Write an index cache file to disk.

    :param index: Issue index to serialize.
    :type index: IssueIndex
    :param cache_path: Path to cache file.
    :type cache_path: Path
    :param file_mtimes: File modification time mapping.
    :type file_mtimes: Dict[str, float]
    """
    cache = IndexCache(
        version=1,
        built_at=datetime.now(timezone.utc),
        file_mtimes=file_mtimes,
        issues=list(index.by_id.values()),
        reverse_deps={
            target: [issue.identifier for issue in issues]
            for target, issues in index.reverse_dependencies.items()
        },
    )
    payload = {
        "version": cache.version,
        "built_at": cache.built_at.isoformat().replace("+00:00", "Z"),
        "file_mtimes": cache.file_mtimes,
        "issues": [
            issue.model_dump(by_alias=True, mode="json") for issue in cache.issues
        ],
        "reverse_deps": cache.reverse_deps,
    }
    cache_path.parent.mkdir(parents=True, exist_ok=True)
    cache_path.write_text(
        json.dumps(payload, indent=2, sort_keys=False),
        encoding="utf-8",
    )


def build_index_from_cache(
    issues: List[IssueData], reverse_deps: Dict[str, List[str]]
) -> IssueIndex:
    """Rebuild an IssueIndex from cached data.

    :param issues: Issues from cache.
    :type issues: List[IssueData]
    :param reverse_deps: Reverse dependency mapping.
    :type reverse_deps: Dict[str, List[str]]
    :return: IssueIndex populated from cache.
    :rtype: IssueIndex
    """
    index = IssueIndex()
    for issue in issues:
        index.by_id[issue.identifier] = issue
        index.by_status.setdefault(issue.status, []).append(issue)
        index.by_type.setdefault(issue.issue_type, []).append(issue)
        if issue.parent is not None:
            index.by_parent.setdefault(issue.parent, []).append(issue)
        for label in issue.labels:
            index.by_label.setdefault(label, []).append(issue)

    for target, issue_ids in reverse_deps.items():
        index.reverse_dependencies[target] = [
            index.by_id[identifier]
            for identifier in issue_ids
            if identifier in index.by_id
        ]

    return index

File: rust/src/cache.rs
//! Index cache utilities for Kanbus.

use std::collections::BTreeMap;
use std::path::Path;
use std::sync::Arc;

use chrono::{DateTime, SecondsFormat, Utc};
use serde::{Deserialize, Serialize};

use crate::error::KanbusError;
use crate::index::IssueIndex;
use crate::models::IssueData;

/// Serialized cache representation for the issue index.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IndexCache {
    pub version: u32,
    pub built_at: DateTime<Utc>,
    pub file_mtimes: BTreeMap<String, f64>,
    pub issues: Vec<IssueData>,
    pub reverse_deps: BTreeMap<String, Vec<String>>,
}

/// Collect file modification times for issues.
pub fn collect_issue_file_mtimes(
    issues_directory: &Path,
) -> Result<BTreeMap<String, f64>, KanbusError> {
    let mut mtimes = BTreeMap::new();
    let entries =
        std::fs::read_dir(issues_directory).map_err(|error| KanbusError::Io(error.to_string()))?;
    for entry in entries {
        let entry = entry.map_err(|error| KanbusError::Io(error.to_string()))?;
        let path = entry.path();
        if path.extension().and_then(|ext| ext.to_str()) != Some("json") {
            continue;
        }
        let mtime = mtime_from_entry(&entry)?;
        if let Some(name) = entry.file_name().to_str() {
            mtimes.insert(name.to_string(), mtime);
        }
    }
    Ok(mtimes)
}

fn mtime_from_entry(entry: &std::fs::DirEntry) -> Result<f64, KanbusError> {
    let metadata = entry
        .metadata()
        .map_err(|error| KanbusError::Io(error.to_string()))?;
    let modified = metadata
        .modified()
        .map_err(|error| KanbusError::Io(error.to_string()))?;
    let duration = modified
        .duration_since(std::time::UNIX_EPOCH)
        .map_err(|error| KanbusError::Io(error.to_string()))?;
    Ok(normalize_mtime(duration.as_secs_f64()))
}

fn normalize_mtime(value: f64) -> f64 {
    (value * 1_000_000.0).round() / 1_000_000.0
}

/// Load cached index if the cache is valid.
pub fn load_cache_if_valid(
    cache_path: &Path,
    issues_directory: &Path,
) -> Result<Option<IssueIndex>, KanbusError> {
    if !cache_path.exists() {
        return Ok(None);
    }
    let contents =
        std::fs::read_to_string(cache_path).map_err(|error| KanbusError::Io(error.to_string()))?;
    let payload: serde_json::Value =
        serde_json::from_str(&contents).map_err(|error| KanbusError::Io(error.to_string()))?;
    let file_mtimes: BTreeMap<String, f64> = serde_json::from_value(
        payload
            .get("file_mtimes")
            .cloned()
            .unwrap_or_else(|| serde_json::json!({})),
    )
    .map_err(|error| KanbusError::Io(error.to_string()))?;
    let current_mtimes = collect_issue_file_mtimes(issues_directory)?;
    if file_mtimes != current_mtimes {
        return Ok(None);
    }
    let issues: Vec<IssueData> = serde_json::from_value(
        payload
            .get("issues")
            .cloned()
            .unwrap_or_else(|| serde_json::json!([])),
    )
    .map_err(|error| KanbusError::Io(error.to_string()))?;
    let reverse_deps: BTreeMap<String, Vec<String>> = serde_json::from_value(
        payload
            .get("reverse_deps")
            .cloned()
            .unwrap_or_else(|| serde_json::json!({})),
    )
    .map_err(|error| KanbusError::Io(error.to_string()))?;

    Ok(Some(build_index_from_cache(issues, reverse_deps)))
}

/// Write the index cache to disk.
pub fn write_cache(
    index: &IssueIndex,
    cache_path: &Path,
    file_mtimes: &BTreeMap<String, f64>,
) -> Result<(), KanbusError> {
    let cache = IndexCache {
        version: 1,
        built_at: Utc::now(),
        file_mtimes: file_mtimes.clone(),
        issues: index
            .by_id
            .values()
            .map(|issue| issue.as_ref().clone())
            .collect(),
        reverse_deps: index
            .reverse_dependencies
            .iter()
            .map(|(target, issues)| {
                (
                    target.clone(),
                    issues
                        .iter()
                        .map(|issue| issue.identifier.clone())
                        .collect(),
                )
            })
            .collect(),
    };
    let payload = serde_json::json!({
        "version": cache.version,
        "built_at": cache.built_at.to_rfc3339_opts(SecondsFormat::Secs, true),
        "file_mtimes": cache.file_mtimes,
        "issues": cache.issues,
        "reverse_deps": cache.reverse_deps,
    });
    if let Some(parent) = cache_path.parent() {
        std::fs::create_dir_all(parent).map_err(|error| KanbusError::Io(error.to_string()))?;
    }
    std::fs::write(
        cache_path,
        serde_json::to_string_pretty(&payload)
            .map_err(|error| KanbusError::Io(error.to_string()))?,
    )
    .map_err(|error| KanbusError::Io(error.to_string()))?;
    Ok(())
}

/// Rebuild an IssueIndex from cached data.
pub fn build_index_from_cache(
    issues: Vec<IssueData>,
    reverse_deps: BTreeMap<String, Vec<String>>,
) -> IssueIndex {
    let mut index = IssueIndex::new();
    for issue in issues {
        let shared = Arc::new(issue);
        index
            .by_id
            .insert(shared.identifier.clone(), Arc::clone(&shared));
        index
            .by_status
            .entry(shared.status.clone())
            .or_default()
            .push(Arc::clone(&shared));
        index
            .by_type
            .entry(shared.issue_type.clone())
            .or_default()
            .push(Arc::clone(&shared));
        if let Some(parent) = shared.parent.clone() {
            index
                .by_parent
                .entry(parent)
                .or_default()
                .push(Arc::clone(&shared));
        }
        for label in &shared.labels {
            index
                .by_label
                .entry(label.clone())
                .or_default()
                .push(Arc::clone(&shared));
        }
    }
    for (target, ids) in reverse_deps {
        let mut issues = Vec::new();
        for identifier in ids {
            if let Some(issue) = index.by_id.get(&identifier) {
                issues.push(Arc::clone(issue));
            }
        }
        index.reverse_dependencies.insert(target, issues);
    }
    index
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Utc;
    use tempfile::tempdir;

    fn sample_issue(identifier: &str) -> IssueData {
        IssueData {
            identifier: identifier.to_string(),
            title: "Title".to_string(),
            description: "Desc".to_string(),
            issue_type: "task".to_string(),
            status: "open".to_string(),
            priority: 1,
            assignee: None,
            creator: None,
            parent: None,
            labels: vec!["label".to_string()],
            dependencies: Vec::new(),
            comments: Vec::new(),
            created_at: Utc::now(),
            updated_at: Utc::now(),
            closed_at: None,
            custom: BTreeMap::new(),
        }
    }

    fn sample_index(issue: IssueData) -> IssueIndex {
        let mut index = IssueIndex::new();
        let shared = Arc::new(issue);
        index
            .by_id
            .insert(shared.identifier.clone(), Arc::clone(&shared));
        index
            .by_status
            .entry(shared.status.clone())
            .or_default()
            .push(Arc::clone(&shared));
        index
            .by_type
            .entry(shared.issue_type.clone())
            .or_default()
            .push(Arc::clone(&shared));
        for label in &shared.labels {
            index
                .by_label
                .entry(label.clone())
                .or_default()
                .push(Arc::clone(&shared));
        }
        index
    }

    #[test]
    fn collect_issue_file_mtimes_skips_non_json() {
        let temp = tempdir().unwrap();
        let issues_dir = temp.path().join("issues");
        std::fs::create_dir_all(&issues_dir).unwrap();
        std::fs::write(issues_dir.join("one.json"), "{}").unwrap();
        std::fs::write(issues_dir.join("note.txt"), "skip").unwrap();
        let mtimes = collect_issue_file_mtimes(&issues_dir).unwrap();
        assert_eq!(mtimes.len(), 1);
        assert!(mtimes.contains_key("one.json"));
    }

    #[test]
    fn load_cache_if_valid_round_trips() {
        let temp = tempdir().unwrap();
        let issues_dir = temp.path().join("issues");
        std::fs::create_dir_all(&issues_dir).unwrap();
        std::fs::write(issues_dir.join("issue.json"), "{}").unwrap();
        let mtimes = collect_issue_file_mtimes(&issues_dir).unwrap();

        let cache_path = temp.path().join("cache").join("index.json");
        let issue = sample_issue("kanbus-abc123");
        let index = sample_index(issue);
        write_cache(&index, &cache_path, &mtimes).unwrap();

        let loaded = load_cache_if_valid(&cache_path, &issues_dir).unwrap();
        let loaded = loaded.expect("expected cache hit");
        assert!(loaded.by_id.contains_key("kanbus-abc123"));
    }

    #[test]
    fn load_cache_if_valid_returns_none_on_mtime_change() {
        let temp = tempdir().unwrap();
        let issues_dir = temp.path().join("issues");
        std::fs::create_dir_all(&issues_dir).unwrap();
        let issue_path = issues_dir.join("issue.json");
        std::fs::write(&issue_path, "{}").unwrap();
        let mtimes = collect_issue_file_mtimes(&issues_dir).unwrap();

        let cache_path = temp.path().join("cache").join("index.json");
        let issue = sample_issue("kanbus-abc123");
        let index = sample_index(issue);
        write_cache(&index, &cache_path, &mtimes).unwrap();

        std::fs::write(&issue_path, "{\"updated\": true}").unwrap();
        let loaded = load_cache_if_valid(&cache_path, &issues_dir).unwrap();
        assert!(loaded.is_none());
    }
}
